src/server/frontend/wayland/wl_mir_window.h

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright © 2015-2018 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: Christopher James Halse Rogers <christopher.halse.rogers@canonical.com>
+ */
+
+#ifndef WL_MIR_WINDOW_H
+#define WL_MIR_WINDOW_H
+
+#include "mir/frontend/surface_id.h"
+#include "mir/geometry/size.h"
+
+#include <memory>
+
+struct wl_client;
+struct wl_resource;
+
+namespace mir
+{
+namespace scene
+{
+class SurfaceCreationParameters;
+}
+namespace shell
+{
+class SurfaceSpecification;
+}
+namespace frontend
+{
+class Shell;
+class BasicSurfaceEventSink;
+
+class WlMirWindow
+{
+public:
+    virtual void new_buffer_size(geometry::Size const& buffer_size) = 0;
+    virtual void commit() = 0;
+    virtual void visiblity(bool visible) = 0;
+    virtual void destroy() = 0;
+    virtual ~WlMirWindow() = default;
+} extern * const null_wl_mir_window_ptr;
+
+class WlAbstractMirWindow : public WlMirWindow
```

Comment:
WlAbstractMirWindow is only used in its implementation file and wayland_connector.cpp, so I don't think it needs to be in wl_mir_window.h.

Reply:
I am planning on splitting XdgSurfaceV6 (and later WlShellSurface) into separate files. They will then both need to inherit from WlAbstractMirWindow. I considered making wl_abstract_mir_window.h in addition to wl_mir_window.h (which I can still do if you want, for consistency), but it would be very short and the .cpp would be even shorter.